lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/Int96TimestampValueWriter.java

```diff
@@ -0,0 +1,108 @@
+    private final DateTimeZone timeZone;
+
+    public Int96TimestampValueWriter(ValuesWriter valuesWriter, Type type, PrimitiveType parquetType, DateTimeZone timeZone)
+    {
+        super(parquetType, valuesWriter);
+        this.timestampType = (TimestampType) requireNonNull(type, "type is null");
+        checkArgument(
+                parquetType.getPrimitiveTypeName().equals(PrimitiveType.PrimitiveTypeName.INT96),
+                "parquetType %s is not INT96");
+        this.timeZone = requireNonNull(timeZone, "timeZone is null");
+    }
+
+    @Override
+    public void write(Block block)
+    {
+        int positionCount = block.getPositionCount();
+        long[] localEpochMillis = new long[positionCount];
+        int[] nanosOfMillis = new int[positionCount];
+        int nonNullsCount = 0;
+        if (timestampType.isShort()) {
+            for (int position = 0; position < positionCount; position++) {
+                if (!block.isNull(position)) {
+                    long epochMicros = timestampType.getLong(block, position);
+                    localEpochMillis[nonNullsCount] = floorDiv(epochMicros, MICROSECONDS_PER_MILLISECOND);
+                    nanosOfMillis[nonNullsCount] = floorMod(epochMicros, MICROSECONDS_PER_MILLISECOND) * NANOSECONDS_PER_MICROSECOND;
+                    nonNullsCount++;
+                }
+            }
+        }
+        else {
+            for (int position = 0; position < positionCount; position++) {
+                if (!block.isNull(position)) {
+                    LongTimestamp timestamp = (LongTimestamp) timestampType.getObject(block, position);
+                    long epochMicros = timestamp.getEpochMicros();
+                    localEpochMillis[nonNullsCount] = floorDiv(epochMicros, MICROSECONDS_PER_MILLISECOND);
+                    int nanosOfMicro = timestamp.getPicosOfMicro() / PICOSECONDS_PER_NANOSECOND;
+                    nanosOfMillis[nonNullsCount] = (floorMod(epochMicros, MICROSECONDS_PER_MILLISECOND) * NANOSECONDS_PER_MICROSECOND) + nanosOfMicro;
+                    nonNullsCount++;
+                }
+            }
+        }
+
+        for (int i = 0; i < nonNullsCount; i++) {
+            long epochMillis = timeZone.convertLocalToUTC(localEpochMillis[i], false);
+            long epochDay = floorDiv(epochMillis, MILLISECONDS_PER_DAY);
+            int julianDay = JULIAN_EPOCH_OFFSET_DAYS + toIntExact(epochDay);
+
+            long nanosOfEpochDay = (epochMillis - (epochDay * MILLISECONDS_PER_DAY)) * NANOSECONDS_PER_MILLISECOND;
+            long timeOfDayNanos = nanosOfMillis[i] + nanosOfEpochDay;
+            Binary binary = toBinary(julianDay, timeOfDayNanos);
+            getValueWriter().writeBytes(binary);
+            getStatistics().updateStats(binary);
+        }
+    }
+
+    private static Binary toBinary(int julianDay, long timeOfDayNanos)
+    {
+        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES + Integer.BYTES);
```

Comment:
Looks like new `ByteBuffer` per every value written.
Is it necessary? We seem to be using the buffer only to write the value and `getStatistics().updateStats(binary)` seems to be making an explicit copy.

if this is necessary then maybe `Binary.fromConstantByteArray(..)` would have less overhead (sans `ByteBuffer` object, just `byte[]`).

Thoughts?

Reply:
If we re-use ByteBuffer then we end needing to use `fromReusedByteArray` which causes copy anyway.
I'm switching to `byte[]` though to avoid ByteBuffer object overhead.